File: day21/haskell/day21.hs
import Data.List (elemIndex, splitAt)

data Instruction = 
    Swap Char Char 
    | SwapPos Int Int 
    | Move Int Int 
    | Reverse Int Int 
    | RotatePos Char 
    | RotateLeft Int 
    | RotateRight Int
    deriving (Show)

parse :: [String] -> Instruction
parse ["swap", "position", x, _, _, y]  = SwapPos       (read x) (read y)
parse ["swap", "letter", x, _, _, y]    = Swap          (head x) (head y)
parse ["rotate", "left", x, _]          = RotateLeft    (read x)
parse ["rotate", "right", x, _]         = RotateRight   (read x)
parse ["rotate", _, _, _, _, _, x]      = RotatePos     (head x)
parse ["reverse", _, x, _, y]           = Reverse       (read x) (read y)
parse ["move", _, x, _, _, y]           = Move          (read x) (read y)

exec :: String -> Instruction -> String
exec s (Swap a b)       = map swap s
    where swap c
            | c == a    = b
            | c == b    = a
            | otherwise = c
exec s (SwapPos x y)    = map (snd . swap) . zip [0..] $ s
    where swap t@(i, _)
            | i == x    = (i, s !! y)
            | i == y    = (i, s !! x)
            | otherwise = t
exec s (RotateLeft x)   = drop n s ++ take n s
    where n = x `mod` length s
exec s (RotateRight x)  = exec s (RotateLeft n)
    where n = length s - x
exec s (RotatePos c)    = exec s (RotateRight (f . elemIndex c $ s))
    where f (Just x)
            | x >= 4    = x + 2
            | otherwise = x + 1
exec s (Reverse x y)    = take x s ++ m ++ drop (y + 1) s
    where m = reverse . take (y - x + 1) . drop x $ s
exec s (Move x y)       = c ++ b:d
    where (a, b:bs)   = splitAt x s
          (c, d)        = splitAt y (a ++ bs)

execR :: String -> Instruction -> String
execR s (Swap a b)       = exec s (Swap a b)
execR s (SwapPos x y)    = exec s (SwapPos x y)
execR s (RotateLeft x)   = exec s (RotateRight x)
execR s (RotateRight x)  = exec s (RotateLeft x)
execR s (Reverse x y)    = exec s (Reverse x y)
execR s (Move x y)       = exec s (Move y x)
execR s (RotatePos c)    = head $ filter f [exec s (RotateLeft x) | x <- [1..]]
    where f = (== s) . flip exec (RotatePos c)

run :: (String -> Instruction -> String) -> String -> [String] -> String
run f s = foldl f s . map (parse . words)

main :: IO ()
main = do
    input <- lines <$> readFile "../input.txt"
    print $ run exec  "abcdefgh" input
    print $ run execR "fbgdceah" (reverse input)

File: day21/python/day21.py
from re import compile
from itertools import chain

swapPosRegex = compile(r'swap position (\d) with position (\d)')
swapLetterRegex = compile(r'swap letter (\w) with letter (\w)')
rotateRegex = compile(r'rotate (left|right) (\d) steps?')
rotatePosRegex = compile(r'rotate based on position of letter (\w)')
reverseRegex = compile(r'reverse positions (\d) through (\d)')
moveRegex = compile(r'move position (\d) to position (\d)')

def swapPos(pword, x, y):
    pword[x], pword[y] = pword[y], pword[x]
    return pword

def swapLetter(pword, x, y):
    return swapPos(pword, pword.index(x), pword.index(y))

def rotate(pword, dir, x):
    l = len(pword)
    return pword[(dir * x) % l:] + pword[:(dir * x) % l]

def rotatePos(pword, x):
    n = pword.index(x)
    return rotate(pword, -1, 1 + n + (0 if n < 4 else 1))

def reverse(pword, x, y):
    return list(chain(pword[:x], reversed(pword[x:y + 1]), pword[y + 1:]))

def move(pword, x, y):
    z = pword.pop(x)
    pword.insert(y, z)
    return pword

def revRotatePos(pword, x):
    pw = pword[::]
    while rotatePos(pw, x) != pword:
        pw = rotate(pw, 1, 1)
    return pw

def part1(password, input):
    password = list(password)
    for ins in input:
        if ins.startswith("swap position"):
            match = swapPosRegex.match(ins)
            password = swapPos(password, *map(int, match.groups()))
        elif ins.startswith("swap letter"):
            match = swapLetterRegex.match(ins)
            password = swapLetter(password, *match.groups())
        elif ins.startswith("rotate based"):
            match = rotatePosRegex.match(ins)
            password = rotatePos(password, *match.groups())
        elif ins.startswith("rotate"):
            d, s = rotateRegex.match(ins).groups()
            password = rotate(password, (1 if d == 'left' else -1), int(s))
        elif ins.startswith("reverse"):
            match = reverseRegex.match(ins)
            password = reverse(password, *map(int, match.groups()))
        elif ins.startswith("move"):
            match = moveRegex.match(ins)
            password = move(password, *map(int, match.groups()))
    return ''.join(password)

def part2(password, input):
    password = list(password)
    for ins in reversed(input):
        if ins.startswith("swap position"):
            match = swapPosRegex.match(ins)
            password = swapPos(password, *map(int, match.groups()))
        elif ins.startswith("swap letter"):
            match = swapLetterRegex.match(ins)
            password = swapLetter(password, *match.groups())
        elif ins.startswith("rotate based"):
            match = rotatePosRegex.match(ins)
            password = revRotatePos(password, *match.groups())
        elif ins.startswith("rotate"):
            d, s = rotateRegex.match(ins).groups()
            password = rotate(password, (-1 if d == 'left' else 1), int(s))
        elif ins.startswith("reverse"):
            match = reverseRegex.match(ins)
            password = reverse(password, *map(int, match.groups()))
        elif ins.startswith("move"):
            match = moveRegex.match(ins)
            password = move(password, *reversed(list(map(int, match.groups()))))
    return ''.join(password)

input = open("../input.txt").read()
input = [x.strip() for x in input.split("\n")]
print(part1("abcdefgh", input))
print(part2("fbgdceah", input))